DataFormats/simulation/include/SimulationDataFormat/MCCompLabel.h

```diff
@@ -47,54 +49,84 @@ class MCCompLabel
-  bool isPosTrackID() const { return getTrackID() >= 0; }
-  // return 1 if the tracks are the same and have labels>=0,
-  // 0 if the tracks are the same but at least one of them has label<0
+  // check if label corresponds to real particle
+  bool isValid() const { return mLabel != Invalid; }
+  // check if label was assigned as for correctly identified particle
+  bool isSetAndValid() const { return isSet() && isValid(); }
+
+  // check if label was assigned as for incorrectly identified particle
+  bool isFake() const { return mLabel & Fake; }
+  // check if label was assigned as for correctly identified particle
+  bool isCorrect() const { return !isFake() && isSetAndValid(); }
+
+  // return 1 if the tracks are the same and correctly identified
+  // 0 if the tracks are the same but at least one of them is fake
   // -1 otherwhise
   int compare(const MCCompLabel& other) const
   {
     if (getEventID() != other.getEventID() || getSourceID() != other.getSourceID()) {
       return -1;
     }
     int tr1 = getTrackID(), tr2 = other.getTrackID();
-    return (tr1 == tr2) ? 1 : ((tr1 == -tr2) ? 0 : -1);
+    return (tr1 == tr2) ? ((isCorrect() && other.isCorrect()) ? 1 : 0) : -1;
   }
 
   // conversion operator
   operator ULong64_t() const { return mLabel; }
   // allow to retrieve bare label
   ULong64_t getRawValue() const { return mLabel; }
 
-  // comparison operator, compares only label, not evential weight info
+  // comparison operator, compares only label, not evential weight or correctness info
   bool operator==(const MCCompLabel& other) const { return (mLabel & maskFull) == (other.mLabel & maskFull); }
   // invalidate
   void unset() { mLabel = NotSet; }
-  void set(int trackID, int evID, int srcID)
+  void invalidate() { mLabel = Invalid; }
+  void setFakeFlag(bool v = true)
+  {
+    if (v) {
+      mLabel |= Fake;
+    } else {
+      mLabel &= ~Fake;
+    }
+  }
+
+  void set(unsigned int trackID, int evID, int srcID, bool fake)
   {
     /// compose label: the track 1st cast to UInt32_t to preserve the sign!
-    mLabel = (maskTrackID & static_cast<ULong64_t>(static_cast<UInt_t>(trackID))) |
+    mLabel = (maskTrackID & static_cast<ULong64_t>(trackID)) |
              (maskEvID & static_cast<ULong64_t>(evID)) << nbitsTrackID |
              (maskSrcID & static_cast<ULong64_t>(srcID)) << (nbitsTrackID + nbitsEvID);
+    if (fake) {
+      setFakeFlag();
+    }
   }
 
   int getTrackID() const { return static_cast<int>(mLabel & maskTrackID); }
+  int getTrackIDSigned() const { return isFake() ? -getTrackID() : getTrackID(); }
   int getEventID() const { return (mLabel >> nbitsTrackID) & maskEvID; }
   int getSourceID() const { return (mLabel >> (nbitsTrackID + nbitsEvID)) & maskSrcID; }
-  void get(int& trackID, int& evID, int& srcID)
+  void get(int& trackID, int& evID, int& srcID, bool& fake)
   {
     /// parse label
-    trackID = getTrackID();
+    trackID = getTrackIDSigned();
```

Comment:
shouldn't this be getTrackID, since we also return the fake flag?

Reply:
yes